From f5c984699c50d2150e14d7e4f21ed99e837a78c8 Mon Sep 17 00:00:00 2001
From: "yi.zhang1" <yi.zhang1@amlogic.com>
Date: Tue, 21 Nov 2023 14:28:59 +0800
Subject: [PATCH 7/9] transcoding: ctrl C quit not correct [1/1]

PD#SWPL-146837

Problem:
transcoding use ctrl C quit, flow not correct.

Solution:
modify the quit flow.

Verify:
local

Signed-off-by: yi.zhang1 <yi.zhang1@amlogic.com>
Change-Id: I04b4c97362a5da9d907e680c62846c2f10f32f6f
---
 gst-aml-videotranscoding/video_transcoding.c  | 5 ++---
 videotranscoding-demo/videotranscoding_demo.c | 6 +++++-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/gst-aml-videotranscoding/video_transcoding.c b/gst-aml-videotranscoding/video_transcoding.c
index 761a347..322a3d5 100644
--- a/gst-aml-videotranscoding/video_transcoding.c
+++ b/gst-aml-videotranscoding/video_transcoding.c
@@ -242,14 +242,13 @@ void video_transcoding_deinit(HANDLE *handle){
     /* free resources */
     data->init = FALSE;
     data->playing = FALSE;
+    g_main_loop_quit(data->main_loop);
     if (pthread_join(data->tid, NULL) ==0) {
         printf("workloop thread exited successful\n");
     } else {
         printf("workloop thread exited fail\n");
     }
-
-    //g_main_loop_quit(data->main_loop);
-    //g_main_loop_unref(data->main_loop);
+    g_main_loop_unref(data->main_loop);
     gst_object_unref(data->pipeline);
     free(data);
 
diff --git a/videotranscoding-demo/videotranscoding_demo.c b/videotranscoding-demo/videotranscoding_demo.c
index a7c8bf1..2c765d8 100644
--- a/videotranscoding-demo/videotranscoding_demo.c
+++ b/videotranscoding-demo/videotranscoding_demo.c
@@ -127,8 +127,12 @@ static void SignalHandler(int signum){
     if ( NULL == handle ) {
         return;
     }
+    CustomData_App *data = (CustomData_App *)handle_app;
     video_transcoding_stop(handle);
     video_transcoding_deinit(handle);
+    gst_element_set_state(data->pipeline, GST_STATE_NULL);
+    gst_object_unref(data->pipeline);
+    free(data);
     exit(1);
 }
 
@@ -394,7 +398,7 @@ int main(int argc, char *argv[]) {
     /* release source */
     gst_element_set_state(data->pipeline, GST_STATE_NULL);
     gst_object_unref(data->pipeline);
-
+    free(data);
     return 0;
 }
 
-- 
2.25.1

